### ISSUE TITLE
Expose string<->byte conversions

### ISSUE THREAD

**Issue by gambiteer** [gambit/gambit#298]
I defined the following functions to convert between UTF8 strings and u8vectors for the R7RS byte-vector to string benchmark:
```
(define (string->utf8 s)
  (with-output-to-u8vector
   (list char-encoding: 'UTF-8)
   (lambda ()
     (display s))))

(define (utf8->string v)
  (let ((p (open-input-u8vector (list char-encoding: 'UTF-8 init: v))))
    (list->string (read-all p read-char))))
```
I believe they're correct ;-), but perhaps there should be functions
```
(string->u8vector s [encoding])
(u8vector->string v [encoding])
```
that expose the underlying routines in `_io.scm` without going through Gambit's I/O system.

**Comment by vyzo** [comment]
A potentially more efficient implementation of the byte->string conversion, as defined in gerbil:
```
(define (bytes->string bstr #!optional (enc 'UTF-8))
  (let* ((in (open-input-u8vector `(char-encoding: ,enc init: ,bstr)))
         (len (u8vector-length bstr))
         (out (make-string len))
         (n (read-substring out 0 len in)))
    (string-shrink! out n)
    out))
```

I agree though, we should have native byte<->string conversion functions that elide the i/o subsystem.

**Comment by gambiteer** [comment]
What do you use in gerbil for `string->bytes`?

**Comment by gambiteer** [comment]
Thanks, that program makes the benchmark run more than twice as fast:
```
+!CSVLINE!+gambitc-v4.8.8,bv2string:1000:1000:100,2.36777925491333
```
versus
```
+!CSVLINE!+gambitc-v4.8.8,bv2string:1000:1000:100,5.416170358657837
```

**Comment by vyzo** [comment]
`string->bytes` is defined like this:
```
(define (string->bytes str #!optional (enc 'UTF-8))
  (substring->bytes str 0 (string-length str) enc))

(define (substring->bytes str start end #!optional (enc 'UTF-8))
  (let ((out (open-output-u8vector `(char-encoding: ,enc))))
    (write-substring str start end out)
    (get-output-u8vector out)))
```

This might be a little faster than display, but not dramatically so.

**Comment by vyzo** [comment]
I have also implemented direct `string<->utf8` conversions in Gerbil stdlib, that don't use ports at all here:
https://github.com/vyzo/gerbil/blob/master/src/std/text/utf8.ss

I can provide a compiled to pure gambit version if you want to benchmark, it doesn't have any dependencies other than `raise-io-error`, which can simply become `error`.

**Comment by feeley** [comment]
I think the performance of your code can be improved by not having to compute the length of the result before doing the conversion (for string->utf8), or the string-shrink! (for utf8->string).  There's also the issue that if a thread modifies the string after the u8vector length is computed, and before doing the conversion, you might index the u8vector out of bound. Look at the implementation of read-chunk in read-line in _io.scm (note that "loop" is not tail-recursive). Some benchmarking would have to be done to see which approach is fastest and by how much.

**Comment by vyzo** [comment]
> I think the performance of your code can be improved by not having to compute the length of the result before doing the conversion (for string->utf8), or the string-shrink! (for utf8->string).

Good point marc; we can allocate a 4x u8vector in string->utf8 and return the length in utf8-encode! and then shrink. utf8->string already avoids counting. 

> There's also the issue that if a thread modifies the string after the u8vector length is computed, and before doing the conversion, you might index the u8vector out of bound.

Well, I tend to think of such concurrency issues as user, not library errors. But if it can be cheaply avoided, it should be.

**Comment by feeley** [comment]
Please read the code of read-line in _io.scm which shows another approach which allocates a u8vector of the correct size (no shrinking necessary).  The recursion however may add a cost... benchmark to see.

**Comment by feeley** [comment]
The following non-tail recursive version is about twice as fast as your two-pass version.

```
(define (rec-string->utf8 str)
  (let lp ((i 0) (j 0))
    (if (##fx< i (##string-length str))
        (let* ((char (##string-ref str i))
               (c (##char->integer char)))
          (cond
           ((##fx<= c #x7f)
            (let ((u8v (lp (##fx+ i 1) (##fx+ j 1))))
              (##u8vector-set! u8v j c)
              u8v))
           ((##fx<= c #x7ff)
            (let ((u8v (lp (##fx+ i 1) (##fx+ j 2))))
              (##u8vector-set! u8v j (##fxior #xc0 (##fxarithmetic-shift-right c 6)))
              (##u8vector-set! u8v (##fx+ j 1) (##fxior #x80 (##fxand c #x3f)))
              u8v))
           ((##fx<= c #xffff)
            (let ((u8v (lp (##fx+ i 1) (##fx+ j 3))))
              (##u8vector-set! u8v j (##fxior #xe0 (##fxarithmetic-shift-right c 12)))
              (##u8vector-set! u8v (##fx+ j 1) (##fxior #x80 (##fxand (##fxarithmetic-shift-right c 6) #x3f)))
              (##u8vector-set! u8v (##fx+ j 2) (##fxior #x80 (##fxand c #x3f)))
              u8v))
           (else                              ; max char is #x10ffff
            (let ((u8v (lp (##fx+ i 1) (##fx+ j 4))))
              (##u8vector-set! u8v j (##fxior #xf0 (##fxarithmetic-shift-right c 18)))
              (##u8vector-set! u8v (##fx+ j 1) (##fxior #x80 (##fxand (##fxarithmetic-shift-right c 12) #x3f)))
              (##u8vector-set! u8v (##fx+ j 2) (##fxior #x80 (##fxand (##fxarithmetic-shift-right c 6) #x3f)))
              (##u8vector-set! u8v (##fx+ j 3) (##fxior #x80 (##fxand c #x3f)))
              u8v))))
        (##make-u8vector j))))
```

**Comment by gambiteer** [comment]
Sweet.

**Comment by vyzo** [comment]
Interesting and very impressive, but isn't the non-tail recursion problematic for long strings?
I will also try against a single pass version that uses a 4x u8vector + shrink.

**Comment by vyzo** [comment]
So the single pass version is a faster than the two-pass version, but not 2x.
Marc, can you try the benchmark with the updated 1-pass version I sent?

**Comment by vyzo** [comment]
Also, do you want me to prepare a patch for whichever version we decide to use?
It would be nice to have fast `string<->utf8` conversions natively, for all Gambit programs.

**Comment by feeley** [comment]
Yes the non-tail recursive version may have worse performance if the string is long (overflows of the stack section can slow things down).  But since it is likely that typically strings are short (say less than 100 characters) it is good to use this approach for short strings.  When strings are long the two pass version can be used or the string-shrink version (there's a space/time tradeoff here that has to be thought through).  Note that the two pass version can be improved by inlining the utf8 character length function.

A patch for Gambit will be nice.

**Comment by vyzo** [comment]
We can use a 3-level decision:
- if the string is small (say 100 chars or so), use the non-tail recursive version
- if the string is medium (say some X Kchars or so), then allocate 4x and shrink
- and if the string is large, then count and allocate an exact length u8vector

**Comment by feeley** [comment]
That's probably overkill... the size of the code is also a consideration, particularly for the universal backend. In fact, if the non-tail recursive version is marginally better for small strings (I haven't benchmarked) it might be better to have a single approach.  Of course we could also implement a switch, similar to (use-fast-bignum-algorithms), to enable the 3-level implementation, and that could be the default with the C backend.  Careful benchmarking will help make a good decision and determine what are the best definitions of "small", "medium" and "large" strings.
